components/application-mgt/org.wso2.carbon.identity.application.common/src/main/java/org/wso2/carbon/identity/application/common/ApplicationAuthenticatorService.java

```diff
@@ -121,4 +134,73 @@ public void removeRequestPathAuthenticator(RequestPathAuthenticatorConfig authen
         }
     }
+
+    public LocalAuthenticatorConfig createUserDefinedLocalAuthenticator(LocalAuthenticatorConfig authenticatorConfig,
+              AuthenticatorType type, String tenantDomain) throws AuthenticatorMgtException {
+
+        LocalAuthenticatorConfig config = getLocalAuthenticatorByName(authenticatorConfig.getName());
+        if (config != null) {
+            ErrorMessages error = ErrorMessages.ERROR_AUTHENTICATOR_ALREADY_EXIST;
+            throw new AuthenticatorMgtClientException(error.getCode(), error.getMessage(),
+                    String.format(error.getDescription(), authenticatorConfig.getName()));
+        }
+        validateAuthenticatorDefinedByType(authenticatorConfig);
+
+        LocalAuthenticatorConfig createdConfig = CACHE_BACKED_DAO.addUserDefinedLocalAuthenticator(
+                authenticatorConfig, IdentityTenantUtil.getTenantId(tenantDomain), type);
+
+        localAuthenticators.add(createdConfig);
+        return createdConfig;
+    }
+
+    public LocalAuthenticatorConfig updateUserDefinedLocalAuthenticator(LocalAuthenticatorConfig authenticatorConfig,
+            String tenantDomain) throws AuthenticatorMgtException {
+
+        LocalAuthenticatorConfig existingAuthenticatorConfig = getExistingAuthenticatorConfig(
+                authenticatorConfig.getName());
+
+        LocalAuthenticatorConfig updatedConfig = CACHE_BACKED_DAO.updateUserDefinedLocalAuthenticator(
+                existingAuthenticatorConfig, authenticatorConfig, IdentityTenantUtil.getTenantId(tenantDomain));
+
+        localAuthenticators.remove(existingAuthenticatorConfig);
+        localAuthenticators.add(updatedConfig);
+        return updatedConfig;
+    }
+
+    public void deleteUserDefinedLocalAuthenticator(String authenticatorName, String tenantDomain)
+            throws AuthenticatorMgtException {
+
+        LocalAuthenticatorConfig authenticatorConfig = getExistingAuthenticatorConfig(authenticatorName);
+
+        CACHE_BACKED_DAO.deleteUserDefinedLocalAuthenticator(authenticatorName,
+                IdentityTenantUtil.getTenantId(tenantDomain));
+        localAuthenticators.remove(authenticatorConfig);
+    }
+
+    private LocalAuthenticatorConfig getExistingAuthenticatorConfig(String authenticatorName)
+            throws AuthenticatorMgtClientException {
+
+        LocalAuthenticatorConfig authenticatorConfig = getLocalAuthenticatorByName(authenticatorName);
+        if (authenticatorConfig == null) {
+            ErrorMessages error = ErrorMessages.ERROR_NOT_FOUND_AUTHENTICATOR;
+            throw new AuthenticatorMgtClientException(error.getCode(), error.getMessage(),
+                    String.format(error.getDescription(), authenticatorName));
+        }
+        validateAuthenticatorDefinedByType(authenticatorConfig);
+
+        return authenticatorConfig;
+
+    }
+
+    private void validateAuthenticatorDefinedByType(LocalAuthenticatorConfig authenticatorConfig)
```

Comment:
I think we can just say validateDefinedByType

Reply:
Addressed